### ISSUE TITLE
Documentation is not valid for Nginx - Set up Nginx proxy with MinIO Server

### ISSUE THREAD

**Issue by dR3b** [minio/minio#11218]
## Expected Behavior
No Error.

## Current Behavior
Minio CLI Error: The request signature we calculated does not match the signature you provided. Check your key and signing method.

## Possible Solution
https://docs.min.io/docs/setup-nginx-proxy-with-minio.html

Nginx:
```
location / {
  proxy_set_header Host $host;
}
```
use `$host` instead of `$http_host`.

## Context
The example documentation is not valid for Nginx. 

https://github.com/yandex/gixy/blob/master/docs/en/plugins/hostspoofing.md

**Comment by ravindk89** [comment]
Thanks for reporting this. At this time the focus for our product documentation is on the MinIO Server and Kubernetes offerings, followed closely by our other first-party products (MinIO Console, MinIO KES, MinIO KES, etc.) 

I intend to eventually come back around to the Cookbook docs, but it depends heavily on the amount of time and resources available alongside other competing priorities. If the community  can validate the proposed solution as generally applicable, I would be happy to accept a PR for this.

**Comment by rluetzner** [comment]
I am also having trouble with the invalid nginx documentation. I managed to setup nginx as a reverse proxy for MinIO, but the part about "Proxy requests based on the bucket" in https://docs.min.io/docs/setup-nginx-proxy-with-minio.html seems to be plain wrong. I spent two days trying to get this to work and I feel that that is not even possible any longer. None of the other proxy setups contain any information on how to do this.

The weird thing is that I don't get any errors, but instead a bucket listing simply won't return any object. The same request made directly to MinIO returns the expected objects.

I'm not quite sure if this is working as intended, but if routing based on buckets is indeed no (longer?) possible with reverse proxys, please remove this information from the documentation.

**Comment by dR3b** [comment]
Why can't you just fix this error quickly? Finding such an error is really tedious. Thanks.

**Comment by ravindk89** [comment]
As noted in my original comment, if the community can validate and explore your proposed solution and verify that it works *at least as generally* as the current example, then I can merge it in. In absence of that, my current plan is to *remove* the majority of cookbook content which pertain to third-party applications. We may selectively publish specific procedures once we are in a position to properly maintain and support the related third-party application, specifically to avoid situations like this where we are adopting a support burden for a non-MinIO application. Right now, we do not have the combination of time, resources, and subject matter expertise to reasonably maintain this content alongside our core products. 

To be more specific as to why I do not think this is as "quick" of a fix as implied:

Both `$host` and `$http_host` are valid NGINX configuration variables with specific behaviors, and therefore they have specific effect on the proxy-ed traffic. From the NGINX docs, `$http_host` should match very closely the `$host` value. At least one [SO post](https://stackoverflow.com/questions/15414810/whats-the-difference-of-host-and-http-host-in-nginx) implies `$host` specifically drops the port number, and has some additional defaults if the host header value is unset - this behavior is not explicitly noted in NGINX docs, so I have to assume the SO user is correct. 

Furthermore, other issues opened around the nginx reverse proxy setup do not indicate your solution as generally applicable. Consider issue [8624](https://github.com/minio/minio/issues/8624), where a user was running into issues with their nginx reverse proxy using the suggested `$host` field instead of our documented `$http_host`. [7857](https://github.com/minio/minio/issues/7857) implies that this could have some relation to how NGINX is passing on the host/port information. [7736](https://github.com/minio/minio/issues/7736) also has a user with `$host` that is still running into auth header issues.

Just between those few issues raised, I'd wager that there's some behavior around the hostname/port passed by the proxy server where in certain deployments the signature verification fails. This might even be a clientside problem, as the 7857 issue had problems on one SDK but not on another. Identifying the specific circumstances in which that occurs requires time, resources, and subject matter expertise that I do not have. Other MinIO engineers might have the expertise, but not the time nor the resources. The community is the best resource *at this time* to review the proposed changes and validate that they work in a variety of scenarios.

**Comment by harshavardhana** [comment]
@ravindk89 we should perhaps drop the doc for Nginx and only promote https://github.com/minio/sidekick

Sidekick has many many advantages of Nginx 

- Built with Go TLS stack supports HTTP2.0 proxying
- Can do health-checks without paying for Nginx Plus license 
- Can do intelligent site-level failures in case of buckets with bucket replication on the server-side

**Comment by ravindk89** [comment]
@harshavardhana I've got no objections to that - I don't see an existing example that explicitly captures the "Reverse Proxy" use case (at least for SEO), so if we can allocate some time to explore that, we can drop this tutorial and replace it with a redirect to the Sidekick Reverse Proxy tutorial (tbd).

**Comment by seqizz** [comment]
Just adding a small note: I got caught in the same thing. Since it was an official documentation on minio's website, I thought my setup was wrong or I was doing something incorrect. Original plan was to use bandwidth-limiting features of nginx (and no [trickle does not work](https://github.com/minio/mc/issues/1749)). But even sidekick does not provide such feature, an obviously deprecated documentation should be at least taken offline, imho.

**Comment by rluetzner** [comment]
I also ran into the same problem as described in #8624, when using `$host` instead of `$http_host`. From what I understand `$host` will overwrite the value of the host header if it is already present in the request with the IP or DNS name of the nginx host (depending on how you addressed nginx).

This works if you are sending your requests to nginx directly, but if you insert a load balancer for example, it is vital to keep the initial value of the host header, therefore `$http_host`. The value of the host header is part of the AWS signature calculation and will lead to a `SignatureDoesNotMatch` error if it is not the same value as used in the client request.

I think the main reason why one should use `$host` over `$http_host` is when one is expecting empty header values. These are however mandatory in S3 requests.

For anyone interested I ran into this issue with the following setup:
sidekick (load balancer) --> nginx (reverse proxy) --> minio

The described setup works if I connect to nginx directly, but will fail with `SignatureDoesNotMatch` errors when inserting sidekick.

tl;dr:
Using `$http_host` is safer and more correct than `$http`.

**Comment by ravindk89** [comment]
Closing this ticket - the documentation is valid, but may not cover all possible configurations.

As noted in a previous comment, the current roadmap involves removing all cookbook documentation. We may selectively republish certain cookbook guides if we are confident we can keep them maintained, accurate, and generally applicable. 

Assuming I can keep my own timelines, we should be in place to complete the migration process to our next-gen docs by the end of this month, at which point this documentation will no longer be linked at all. Given a check of our analytics, this page is viewed about 2K times in the last 30 days, but I am not aware of similar issues as the above as a general trend. I would venture therefore that the current documentation is accurate enough for the majority of our audience. 

There might be a middle ground of talking about both values and the circumstances to use them, but I do not have the resources to do that at this time. Others are welcome to open a pull request and assemble information from this issue request and other similar Nginx related issues.
